backend/src/supervisors/supervisor.py

```diff
@@ -1,30 +1,86 @@
 import logging
+from src.prompts import PromptEngine
 from src.router import pick_agent
+from src.agents import DatastoreAgent, GoalAchievedAgent, UnresolvableTaskAgent, MathsAgent
+
+unresolvable_task_agent = UnresolvableTaskAgent()
+goal_achieved_agent = GoalAchievedAgent()
+list_of_agents = [ DatastoreAgent(), MathsAgent(), goal_achieved_agent, unresolvable_task_agent ]
+
+prompt_engine = PromptEngine()
+
+
+def still_solving_tasks(current_agent, iter_count):
+    agent_calling_unresolved = current_agent != "UnresolvableTaskAgent" and current_agent != "GoalAchievedAgent"
+    count_not_exceeded = iter_count < 5
+    return agent_calling_unresolved and count_not_exceeded
```

Comment:
**If we want the agents to be responsible to move on to the next task** we may run the risk of calling the wrong agent that produces a convincing answer. What if we wanted to ask about the users favorite sports team, but instead the router mistakenly chose to search the web and found the generally most popular team? The agent for searching the web would probably be convinced the issue is solved and that we can move on. If we don't verify with the router post completion that the _correct problem_ has been solved via picking the correct agent I think we always run this risk.

**If we want these agents to handle terminating the process**, I'm not sure how this would work. The while loop requires breaking for us to quit the flow either from solving the question (utterance) or abandoning it. If the agent contains the logic it will need to be able to (within it's `invoke()` method) break that loop and return the final value, and I'm not sure of a clean way to put that logic within the agent instead of as a check in the `supervisor.py`